Threat Details: Skeleton with blank Content UiState

ashiagr opened the pull request:
Parent: #13326


### Description

This PR does the following for "Threat Details"
- Creates skeleton vm, fragment, recyclerview, items builder and tests - 2a171eb, 3c0c5dc, 5ae1c65, bfbce02
- Introduces `UiState` with blank `Content` - 30c07fd
- Gets threat by `threatId` using a `UseCase` - 8c1867b, 75e89e2
- Launches `ThreatDetailsActivity` from the `ScanFragment` - c5b7794

### To test

Prerequisite: Make sure both Scan feature flag and MySiteImprovements flag are set to on in the App settings.

1. Select a site with scan capability (and having threats) on the My Site tab
2. Notice that scan menu is displayed
3. Click scan menu item
4. Notice that threats are shown
5. Click a threat item
6. Notice that blank Threat Details screen is displayed

### Merge Instructions

1. Make sure FluxC PR https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1819 is merged and tag is created in develop - Merged
2. Update FluxC hash with tag (created in FluxC develop) - Done in 4f4b947
3. Remove the "Not Ready for Merge label" - Done
4. Merge this PR

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Refactor: Remove unwanted ScanActivity inject

ashiagr committed:
Refactor: Move ScanStateListItemBuilder to builder package

ashiagr committed:
Refactor: Pluralize ScanStateListItemBuilder to ScanStateListItemsBuilder

ashiagr committed:
Introduce ThreatDetailsViewModel

ashiagr committed:
Launch newly created ThreatDetails activity from Scan fragment

ashiagr committed:
Introduce ThreatDetails RecyclerView and Adapter

ashiagr committed:
Refactor: Pass threat id instead of entire model to ShowThreatDetails

ashiagr committed:
Introduce ThreatDetails Activity and Fragment

ashiagr committed:
Introduce GetThreatDetailsUseCase

ashiagr committed:
Introduce blank ThreatDetailsListItemsBuilder

ashiagr committed:
Introduce UiState with blank Content

ashiagr committed:
Update fluxc commit hash

ashiagr committed:
Update UiState with blank content

ashiagr committed:
Fix DiffUtil

ashiagr committed:
Rename use case

ashiagr committed:
Update tests

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/details/adapters/ThreatDetailsAdapter.kt:
```diff
@@ -0,0 +1,71 @@
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): JetpackViewHolder {
+        return when (viewType) {
+            ViewType.ICON.id -> JetpackIconViewHolder(imageManager, parent)
+            ViewType.DESCRIPTION.id -> JetpackDescriptionViewHolder(uiHelpers, parent)
```
These are placeholder ViewTypes, appropriate ones are added in a separate PR #13703.

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/details/ThreatDetailsFragment.kt:
```diff
@@ -0,0 +1,59 @@
+    private fun initViewModel() {
+        viewModel = ViewModelProvider(this, viewModelFactory).get(ThreatDetailsViewModel::class.java)
+        setupObservers()
+        val threatId = requireActivity().intent.getLongExtra(ScanConstants.ARG_THREAT_ID, 0)
```
Idea (💡 ): Why not moving this `ARG_THREAT_ID` constant (in a `companion object`) within the fragment itself and remove the need to create a `ScanConstants` class altogether?

ashiagr replied:
I created `ScanConstants` for keeping `intent` `args` because a similar separate class [ReaderConstants](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderConstants.java#L35) was used to keep them for `Reader` classes. 

No personal preference on this so moved it in 81e165c.

ParaskP7 replied:
Thanks Ashita! 🌟 

I wasn't sure about the intention here and whether this is a pattern you follow elsewhere for some specific reason. It just felt too much to create an extra class instead of keeping this close to its usage.

Maybe I am missing something. Let's keep a note on it and discuss it for next time.

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/details/ThreatDetailsViewModel.kt:
```diff
@@ -0,0 +1,49 @@
+            return
+        }
+        isStarted = true
+        this.threatId = threatId
```
Question (❓): Why is `threadId` exposed as a field, will it be used as such later on? Otherwise, I would suggest just passing the `threadId` parameter to the `getData()` function and remove the `threadId` field for now. Wdyt?

ashiagr replied:
`threatId` will be used later on to perform fix and ignore threat actions (API requires `threatId`) in the future.

ParaskP7 replied:
Great, thanks for the information Ashita! 🌟 

PS: I usually try to avoid exposing fields too soon, later on when it is going to be needed it would be a matter of a single refactor commit, so it doesn't seem as an overhead to me. As such, I usually do that when needed and avoid having the reviewer(s) asking me about any future intentions. Also, sometime this saves me from rolling back since what I expected happening in the future have now changed and I start being sad. 😝

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/details/ThreatDetailsViewModel.kt:
```diff
@@ -0,0 +1,49 @@
+        _uiState.value = buildContentUiState(threatModel)
+    }
+
+    private fun buildContentUiState(model: ThreatModel) = Content(builder.buildThreatDetailsListItems(model))
```
Minor (🔍): The parameter here is named as `model`, while everywhere else it is named as `threadModel`. Consider one or the other. Personally, I would go with plain `model` everywhere since the context in which it is used is `ThreadDetails` anyway. Wdyt?

ashiagr replied:
Done in 11ad474

ParaskP7 commented on WordPress/src/test/java/org/wordpress/android/ui/jetpack/scan/details/ThreatDetailsViewModelTest.kt:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @Test
+    fun `given threat id, then on start, threat details are retrieved`() = test {
```
Minor (🔍): The action of the test sentence probably need to be replaced with `when` (instead of `then`). In the end you might want something like the: `given threat id, when on start is triggered, then threat details are retrieved`

Let me know your thoughts. 🙏  

PS: The same apply for the test below: `given threat id, when on start is triggered, then ui is updated with content`

ashiagr replied:
Oh yes, thanks for noticing 😅. Rephrased in 2811298.

ParaskP7 reviewed (approved):
Great job @ashiagr ! 🌟 

I have tested the solution (as per the instructions) and can verify correctness.

I have left few minors, questions and ideas here and there but overall I am approving the PR since those are not blocking. Feel free to merge the PR anyway. I am not merging the PR myself since I want to give you the opportunity to first check my comments.

PS: I really love the way you split your commits, kudos (❤️)! It really helped me review the PR and understand the way the solution was built. Also, I loved the fact that you added the `Refactor: ` action prefix on your initial commits. Actually, my team used to do that in the past as well and we found out that it works miracles. I wonder if you folks would like to start using this just as well, for a trial period of time and see if you like it. Your thoughts @zwarm @malinajirka ?

ashiagr committed:
Move ARG_THREAT_ID in the fragment's companion object

ashiagr committed:
Rename threatModel to model

ashiagr committed:
Rephrase test sentences

ashiagr committed:
Update fluxc version

ashiagr commented:
Thanks @ParaskP7! I've addressed your comments and answered questions. I've also updated fluxc tag from `develop` and removed "Not Ready for Merge" label. 

Ready for another look 🙏.

ParaskP7 commented:
> Thanks @ParaskP7! I've addressed your comments and answered questions. I've also updated fluxc tag from develop and removed "Not Ready for Merge" label.

Thanks for all the changes @ashiagr! 🌟 

It looks good, I am re-approving and will be merging this PR now.
